### PR TITLE
[C++ API] Add functional overloads for fold, linear, loss, normalization, padding

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29360 Add functional overloads for fold, linear, loss, normalization, padding**
* #29359 Add functional overloads for pixelshuffle, pooling, upsampling, vision
* #29358 Add functional overloads for activation, batchnorm, distance, embedding

This PR adds functional overloads that take the full set of arguments (instead of just Options) for the following functionals:
- fold
- linear
- loss
- normalization
- padding

These new functionals lives in the `torch::nn::functional::detail` namespace and they are only meant to be called from the module forward methods (i.e. they are not public API). This is in preparation for the future change where we make module Options and functional Options two different classes, because if the module forward method has to construct a new functional Options object every time it runs it will be pretty silly and bad performance.

Differential Revision: [D18376975](https://our.internmc.facebook.com/intern/diff/D18376975)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 5f2783f:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=5f2783f4ff18b153914114b6a76995fd9d938a2c).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Review by pbelevich — APPROVED** [review]

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 1]
```diff
@@ -68,16 +91,28 @@ inline Tensor mse_loss(
...
+
+// ============================================================================
 
+namespace detail {
 inline Tensor binary_cross_entropy(
     const Tensor& input,
     const Tensor& target,
-    const BCELossOptions& options = {}) {
-  auto reduction_enum = enumtype::reduction_get_enum(options.reduction());
+    Tensor weight,
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: thanks a lot for the catch! I changed it to pass by const reference

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 2]
```diff
@@ -91,67 +126,116 @@ inline Tensor binary_cross_entropy(
       "!= input nelement (", input.numel(), ")");
   }
 
-  auto weight = options.weight();
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: right it doesn't, although I changed `weight` to pass by const reference just to be safe

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 3]
```diff
@@ -91,67 +126,116 @@ inline Tensor binary_cross_entropy(
...
 inline Tensor multi_margin_loss(
     const Tensor& input,
     const Tensor& target,
-    const MultiMarginLossOptions& options = {}) {
-  TORCH_CHECK(options.p() == 1 || options.p() == 2, "only p == 1 and p == 2 supported");
-  if (options.weight().defined()) {
-    TORCH_CHECK(options.weight().dim() == 1, "weight must be one-dimensional");
+    int64_t p,
+    double margin,
+    Tensor weight,
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: thanks a lot for the catch! I changed it to pass by const reference

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 4]
```diff
@@ -162,105 +246,226 @@ inline Tensor smooth_l1_loss(
...
 }
 
+// ============================================================================
+
+namespace detail {
 inline Tensor multilabel_soft_margin_loss(
     const Tensor& input,
     const Tensor& target,
-    const MultiLabelSoftMarginLossOptions& options = {}) {
+    Tensor weight,
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: thanks a lot for the catch! I changed it to pass by const reference
